File: miscellaneous/tree-bfs/20230410/tree_bfs.py
from __future__ import annotations
from typing import List, Set


def tree_bfs(tree: TreeNode) -> List[int]:
    values: List[int] = []
    seen: Set[int] = set()
    queue: List[TreeNode] = [tree]

    while (queue):
        current = queue.pop(0)
        values.append(current.val)

        if current.val not in seen:
            seen.add(current.val)
            current.left and queue.append(current.left)
            current.right and queue.append(current.right)

    return values


class TreeNode:
    def __init__(self, val: int, left: TreeNode = None, right: TreeNode = None):
        self.val = val
        self.left = left
        self.right = right

File: miscellaneous/tree-bfs/20230410/tree-bfs.php
<?php

declare(strict_types=1);

function treeBfs(TreeNode $tree): array
{
    $queue = [$tree];
    $values = [];

    while ($queue) {
        $current = array_shift($queue);
        $values[] = $current->val;

        $current->left && $queue[] = $current->left;
        $current->right && $queue[] = $current->right;
    }

    return $values;
}


class TreeNode
{
    public function __construct(
        public int $val,
        public ?TreeNode $left = null,
        public ?TreeNode $right = null
    ) {}
}